### PR TITLE
Decimal128 support for Parquet

### PR DESCRIPTION
**Pull request by kuhushukla**
This is not ready for complete review as we need more tests and support for uint64.

### CODE REVIEW

**Comment by sperlingxx** [comment]
It would be nice to add some tests for nested types of decimal128.

**Comment by kuhushukla** [comment]
> It would be nice to add some tests for nested types of decimal128.

Oops. I think I missed a commit. I'm in the process of rebasing. Will update tomorrow with more tests

**Comment by revans2** [comment]
Looks good is there any way to get some parquet files with unsigned longs in them to test that code?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScanBase.scala [hunk 1]
```diff
@@ -796,7 +797,8 @@ trait ParquetPartitionReaderBase extends Logging with Arm with ScanWithMetrics
   }
 
   def needDecimalCast(cv: ColumnView, dt: DataType): Boolean = {
-    cv.getType.isDecimalType && !GpuColumnVector.getNonNestedRapidsType(dt).equals(cv.getType())
+    cv.getType.isDecimalType && !GpuColumnVector.getNonNestedRapidsType(dt).equals(cv.getType()) ||
+      cv.getType.equals(DType.UINT64)
```
Comment (sperlingxx) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScanBase.scala:801]: Could you add some comments to elaborate why `UINT64` needs to be casted to decimal here?

**Review by sperlingxx — DISMISSED** [review]
It looks good generally. 
NIT:  Can you add some comments about unsigned 64 conversion?

**Comment by kuhushukla** [comment]
> Looks good is there any way to get some parquet files with unsigned longs in them to test that code?

Yes that would be nice, isn't it? I will try and make that work.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScanBase.scala (hunk 1 above)
Comment (kuhushukla) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScanBase.scala:801]: Yes I can :) Coming up shortly.

**Review by revans2 — APPROVED** [review]

**Comment by kuhushukla** [comment]
I cant find a simple way to get uint64s on the fly for testing. Can use any ideas I might have missed? I see we had to check in files earlier as well possibly for the same limitation.

**Comment by kuhushukla** [comment]
build

**Comment by revans2** [comment]
FYI:
```
[2021-12-21T16:18:06.489Z] error file=/home/jenkins/agent/workspace/jenkins-rapids_premerge-github-3604/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala message=File line length exceeds 100 characters line=827
```

**Comment by kuhushukla** [comment]
@revans2 Pushed commit to fix the missing piece from verify.

**Comment by kuhushukla** [comment]
build

**Review by revans2 — APPROVED** [review]

**Comment by sperlingxx** [comment]
Hi @kuhushukla, I found this PR also addressed #3475. So, could you add a test case to verify that we are already able to read `UINT_64` column from parquet files ?

**Comment by kuhushukla** [comment]
Adding a parquet file to test UINT64.
```
== Physical Plan ==
GpuColumnarToRow false
+- GpuShuffleCoalesce 2147483647
   +- GpuColumnarExchange gpusinglepartitioning$(), REPARTITION_BY_NUM, [id=#112465]
      +- GpuFileGpuScan parquet [simple_uint64#39706,arr_uint64#39707] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[file:/home/kuhus/Reps/rapids-plugin-4-spark/tests/target/spark320/test-classes/..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<simple_uint64:decimal(20,0),arr_uint64:array<decimal(20,0)>>
```

I created the file using `pyarrow`.

**Comment by kuhushukla** [comment]
@sperlingxx Please take a look.

**Comment by kuhushukla** [comment]
build

**Comment by kuhushukla** [comment]
Looking at the build failure with 32x shims now.

**Comment by jlowe** [comment]
The build failure should be fixed by #4427.

**Comment by sperlingxx** [comment]
LGTM

**Comment by kuhushukla** [comment]
build

**Comment by kuhushukla** [comment]
build

**Comment by kuhushukla** [comment]
build

**Review by revans2 — APPROVED** [review]

**Comment by kuhushukla** [comment]
Running the normal - non databricks premerge build for sanity. Also pinging @sperlingxx for the additional look over and approval.

**Comment by kuhushukla** [comment]
build

**Pull request merged**